protobuf/lib/src/protobuf/extension_registry.dart

```diff
@@ -109,14 +109,22 @@ T _reparseMessage<T extends GeneratedMessage>(
   final messageUnknownFields = message._fieldSet._unknownFields;
   if (messageUnknownFields != null) {
     final codedBufferWriter = CodedBufferWriter();
-    extensionRegistry._extensions[message.info_.qualifiedMessageName]
-        ?.forEach((tagNumber, extension) {
-      final unknownField = messageUnknownFields._fields[tagNumber];
-      if (unknownField != null) {
-        unknownField.writeTo(tagNumber, codedBufferWriter);
-        ensureUnknownFields()._fields.remove(tagNumber);
+
+    if (message is $_MessageSet) {
+      final itemList = messageUnknownFields._fields[_messageSetItemsTag];
+      if (itemList != null) {
+        itemList.writeTo(_messageSetItemsTag, codedBufferWriter);
       }
-    });
+    } else {
+      extensionRegistry._extensions[message.info_.qualifiedMessageName]
+          ?.forEach((tagNumber, extension) {
+        final unknownField = messageUnknownFields._fields[tagNumber];
+        if (unknownField != null) {
+          unknownField.writeTo(tagNumber, codedBufferWriter);
+          ensureUnknownFields()._fields.remove(tagNumber);
+        }
+      });
+    }
```

Comment:
The special case for message sets above is inconsistent with the current handling of unknown fields. Currently if I have unknown fields with tags 1 and 2, and reparse with an extension for 1, the resulting message won't have 2 as an unknown field. The code for message sets OTOH serializes all extensions so 2 will be in unknown fields of the reparsed message.